### ISSUE TITLE
What to do for spoofed user-agents?

### ISSUE THREAD

**Issue by jamesarosen** [polyfillpolyfill/polyfill-service#2240]
This isn't exactly a bug or a feature-request. It's more of a problem (and hopefully the start of a discussion).

We recently switched a site to Polyfill.io and started getting lots of reports of errors like `Promise.all is not a function` and `Promise.resolve is not a function`.

The user-agents are varied, but all for fairly modern browsers (that should definitely have `Promise`):

 * Mobile Safari 12.1.2: `Mozilla/5.0 (iPhone; CPU iPhone OS 12_4_6 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.2 Mobile/15E148 Safari/604.1`
 * Mobile Safari 13.1: `Mozilla/5.0 (iPhone; CPU iPhone OS 13_4_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.1 Mobile/15E148 Safari/604.1`
 * Mobile Safari 13.0.5: `
Mozilla/5.0 (iPhone; CPU iPhone OS 13_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.5 Mobile/15E148 Safari/604.1`
 * Chrome Mobile 81.0.4044: `Mozilla/5.0 (Linux; Android 9; SM-G950U) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Mobile Safari/537.36`

All the IPs come from within Facebook. I assume these are requests from Facebook's crawler with a spoofed user-agent.

On the one hand, you might argue that a spoofed user-agent isn't a problem for Polyfill.io. It's meant to work for "well-behaving clients." On the other hand, (like it or not) Facebook is critical to the success of many businesses.

One possible solution would be to add support for a query-param like `&flags=gated-sync-request` that (1) ships the `detect.js` logic to the client, then (2) requests the necessary polyfills using one or more synchronous XHRs.

**Comment by JakeChampion** [comment]
The Facebook crawler has [two documented user-agents](https://developers.facebook.com/docs/sharing/webmasters/crawler/#identify) that it uses
- facebookexternalhit/1.1 (+http://www.facebook.com/externalhit_uatext.php)
- facebookexternalhit/1.1

That makes me think it is not the Facebook crawler. Have you tried contacting Facebook to see if they know what is causing these requests to your website?

We've been hesitant in the past to implement a feature detection based api like the one you have suggested because of the negative performance impact a snyc xhr has(also sync xhr can be blocked via the [feature policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy/sync-xhr)).

**Comment by jamesarosen** [comment]
> That makes me think it is not the Facebook crawler.

That was my initial thought. When I look at our logs, though, these requests come from exactly the same IPs as `facebookexternalhit` requests. Of course, this could be legitimate users behind a NAT or using a proxy. My current guess is that this isn't the crawler per se, but rather some sort of internal "prerender" service that forwards real user requests on, possibly to show a preview.

> Have you tried contacting Facebook to see if they know what is causing these requests to your website?

I've reached out to some friends who work there. I've looked all over for a way to contact them about technical issues like this but can't find a form anywhere.

> We've been hesitant in the past to implement a feature detection based api like the one you have suggested because of the negative performance impact a snyc xhr has

Absolutely. This is the only idea I've come up with, but I don't love it.
